File: 290. Word Pattern.cpp
class Solution {
public:
    // hash table: O(N)time & space, 38.80%
    bool wordPattern(string pattern, string str) {
        istringstream cut(str);
        vector<string> s;
        string temp;
        while(cut >> temp) s.push_back(temp);
        if(s.size() != pattern.size()) return false;
        unordered_map<char, string> ps;
        unordered_map<string, char> sp;
        for(int i = 0; i < pattern.size(); i++){
            if(!ps.count(pattern[i]) && !sp.count(s[i])){
                ps[pattern[i]] = s[i];
                sp[s[i]] = pattern[i];
            }
            else{
                if(ps[pattern[i]] != s[i] || sp[s[i]] != pattern[i])
                    return false;
            }
        }
        return true;
    }
};

File: 290. Word Pattern.py
class Solution(object):
    
    # hash table: O(N) time & space
    def wordPattern(self, pattern, str):
        """
        :type pattern: str
        :type str: str
        :rtype: bool
        """
        str = str.split()
        if len(pattern) != len(str):
            return False
        p2s, s2p = {}, {}
        for p, s in zip(pattern, str):
            if p not in p2s and s not in s2p:
                p2s[p], s2p[s] = s, p
            elif p in p2s and p2s[p] == s and s in s2p and s2p[s] == p:
                continue 
            else:
                return False
        return True 
                
        
        